lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -1105,42 +1107,49 @@ def configure_hardware_params(self, vm_obj):
             if 'version' in self.params['hardware']:
                 hw_version_check_failed = False
                 temp_version = self.params['hardware'].get('version', 10)
-                try:
-                    temp_version = int(temp_version)
-                except ValueError:
-                    hw_version_check_failed = True
-
-                if temp_version not in range(3, 15):
-                    hw_version_check_failed = True
-
-                if hw_version_check_failed:
-                    self.module.fail_json(msg="Failed to set hardware.version '%s' value as valid"
-                                              " values range from 3 (ESX 2.x) to 14 (ESXi 6.5 and greater)." % temp_version)
-                # Hardware version is denoted as "vmx-10"
-                version = "vmx-%02d" % temp_version
-                self.configspec.version = version
-                if vm_obj is None or self.configspec.version != vm_obj.config.version:
-                    self.change_detected = True
-                if vm_obj is not None:
-                    # VM exists and we need to update the hardware version
-                    current_version = vm_obj.config.version
-                    # current_version = "vmx-10"
-                    version_digit = int(current_version.split("-", 1)[-1])
-                    if temp_version < version_digit:
-                        self.module.fail_json(msg="Current hardware version '%d' which is greater than the specified"
-                                                  " version '%d'. Downgrading hardware version is"
-                                                  " not supported. Please specify version greater"
-                                                  " than the current version." % (version_digit,
-                                                                                  temp_version))
-                    new_version = "vmx-%02d" % temp_version
-                    try:
-                        task = vm_obj.UpgradeVM_Task(new_version)
+                if temp_version == 'latest':
+                    if vm_obj is not None:
```

Comment:
If 'template' parameter is set, and 'version' is also set, then 'vm_obj' is template, it's upgrade is not supported?